### ISSUE TITLE
Daijishou can no longer kill RetoArch when launching a new game

### ISSUE THREAD

**Issue by mattakins** [TapiocaFox/Daijishou#703]
Whenever I have a game loaded in RetroArch and I attempt to launch a different one over it through Daijisho, it takes me to a black screen and I have to force quit RetroArch then launch the game again. When I don't have a game previously loaded, Daijisho will launch the game just fine.

With previous versions of RetroArch, I was able to launch new games over previous games using Daijishou. Something changed in recent RetroArch versions that broke the Daijisho kill feature.

Back in RetroArch 1.16.0 (Oct 2 2023), I was able to launch games from Daijisho over a previously loaded game without closing it first. Daijisho effectively kills the previous game and boots the new one. I do not know which specific version broke this functionality as I only have started using newer RetroArch builds recently.

Thanks for your great work. Let me know if there is any other information I can provide.

**Comment by adriansticoid** [comment]
Would be best to submit logs because something seems to wrong with your setup. I tested with my Retroid Pocket 2S, normally I close content first with a hotkey before starting a new game. This time, I launched a game, pressed the home button launched a different game and it just worked. Did it 5 times and no issues at all.

**Comment by mattakins** [comment]
Thanks for testing. I did some testing and this issue occurs on my Android 13 device but not my Android 12 device. It works fine on that one. I am using the latest versions of Daijishou and the latest nightly of RetroArch on both devices with clean installs of each. How do I export logs from Daijishou?

**Comment by mattakins** [comment]
Upon investigating further user reports on Reddit, I am pretty sure this is a compatibility issue with Android 13.

- https://www.reddit.com/r/EmulationOnAndroid/comments/1bsx03l/retroarch_requiring_manual_close_to_launch/
- https://www.reddit.com/r/RG556/comments/1d7k4f8/anyone_else_experience_this_from_retroarch_and/

Both of these examples report this same issue on Android 13 devices. The first one also reports not experiencing the issue on his Android 12 device. The Retroid Pocket 2S is Android 11, so it would make sense that this issue wouldn't occur on your device.

**Comment by adriansticoid** [comment]
I also have the AYN Odin 2, which is Android 13. I still can't reproduce the issue.

**Comment by mattakins** [comment]
Interesting. I don't think it's my setup since I replicated the issue with a clean install of both Daijishou and RetroArch and because of the other examples of the same issue. I am using a Ayaneo Pocket S.

**Comment by Jetup13** [comment]
I also can't replicate this on any of my devices.

A13/Retroid Pocket 4 Pro
A9/Retroid Pocket 2 Plus
A12/Samsung S10e

Using RetroArch Nightly (c87a88f)
Daijishou version 403 (1.4.71)

Recent commits for RetroArch doesn't show any major change on how RetroArch loads on Android. It sounds like something is either wrong in your setup or the device (similar to how the Odin 2 had major issues).

**Comment by mattakins** [comment]
Thanks for testing. Just had another guy test this on his Pocket S and he experiences this issue as well. Sounds like it's only an issue on certain devices like the Pocket S and RG556. Maybe it's a subset of Android 13 devices as these are the only other reports I can find.

**Comment by Steerwalt** [comment]
Odin 2 Mini here. Same issue.  It doesn’t matter if Retroarch is running a game or not, I have to fully close Retroarch for Daijosho to launch a game on it. Otherwise it crashes.

**Comment by judd43** [comment]
Odin 2 Mini as well here, same issue. I have an original Odin 2 which does not have this issue.

**Comment by tailtwo** [comment]
Same issue on my Retroid Pocket 5 on Android 13, I am using the latest nightly Retroarch.

**Comment by Jetup13** [comment]
Make sure to *properly* quit RetroArch beforehand. You can enable RetroArch to quit when you select close content from quick menu:

> RetroArch > Settings > User Interface > Quit on Close Content > CLI > Save Configuration

There's also various Hotkeys you can enable:

> RetroArch > Settings > Input > Hotkeys 

Pressing Home Button or swiping into Recent Apps beforehand is not considered closing RetroArch *properly*.

**Comment by tailtwo** [comment]
I am pretty sure the reason Daijishou can't kill Retroarch is because it never ask for the permission to do it. I tried to add the permission manually but got :

```
kona:/ $ pm grant com.magneticchen.daijishou android.permission.KILL_BACKGROUND_PROCESSES

Exception occurred while executing 'grant':
java.lang.SecurityException: Permission android.permission.KILL_BACKGROUND_PROCESSES requested by com.magneticchen.daijishou is not a changeable permission type
        at com.android.server.pm.permission.PermissionManagerServiceImpl.grantRuntimePermissionInternal(PermissionManagerServiceImpl.java:1392)
        at com.android.server.pm.permission.PermissionManagerServiceImpl.grantRuntimePermission(PermissionManagerServiceImpl.java:1319)
        at com.android.server.pm.permission.PermissionManagerService.grantRuntimePermission(PermissionManagerService.java:545)
        at android.permission.PermissionManager.grantRuntimePermission(PermissionManager.java:597)
        at com.android.server.pm.PackageManagerShellCommand.runGrantRevokePermission(PackageManagerShellCommand.java:2546)
        at com.android.server.pm.PackageManagerShellCommand.onCommand(PackageManagerShellCommand.java:292)
        at com.android.modules.utils.BasicShellCommandHandler.exec(BasicShellCommandHandler.java:97)
        at android.os.ShellCommand.exec(ShellCommand.java:38)
        at com.android.server.pm.PackageManagerService$IPackageManagerImpl.onShellCommand(PackageManagerService.java:6051)
        at android.os.Binder.shellCommand(Binder.java:1049)
        at android.os.Binder.onTransact(Binder.java:877)
        at android.content.pm.IPackageManager$Stub.onTransact(IPackageManager.java:4313)
        at com.android.server.pm.PackageManagerService$IPackageManagerImpl.onTransact(PackageManagerService.java:6035)
        at android.os.Binder.execTransactInternal(Binder.java:1285)
        at android.os.Binder.execTransact(Binder.java:1244)
```

I guess it should be added in the app manifest so that it is allowed to kill arbritrary processes by using Android APIs.

**Comment by tailtwo** [comment]
Found how to solve the issue for a Retroarch platform, you have to add the `-e QUITFOCUS` argument. Should I submit a PR?

**Comment by h3ckman** [comment]
Also getting this behavior on my Odin Mini.

Tried completely wiping my setup, both Retroarch and Daijisho, and even with a fresh install it still persists.

**Comment by NewbulaCo** [comment]
I've got the same issue on my Retroid Pocket 5.

**Comment by Carter0108** [comment]
Same issue on both my Retroid Pocket 5 with Android 13 and Pixel 7 on Android 15. Daijisho launches the first game no issues but the next game will either be a black screen or sometimes black screen into a not responding popup for Retroarch.

**Comment by mattakins** [comment]
I am also experiencing this on my Retroid Pocket 5.

**Comment by Carter0108** [comment]
> As I've said before, stop using the home button or app switcher and properly quit RetroArch beforehand. Using the home button or app switcher with RetroArch can possibly lead to save data or configuration loss as RetroArch might not get a chance to flush to disk. RetroArch is already a bit funky on Android. This is more of a android issue than a frontend issue.
> 
> RetroArch has many ways you can properly close it.
> 
> I'd suggest the following -
> 
> Enable Quit on Close Content:
> 
> > RetroArch > User Interface > Quit on Close Content > CLI > Save Configuration
> 
> Quick Menu:
> 
> > RetroArch Quick Menu > Close Content
> 
> Hotkeys:
> 
> > RetroArch > Settings > Hotkeys > Quit, Quit (Controller Combo), or Close Content if (Quit on Close Content is enabled) > Save Configuration
> 
> TLDR: Use RetroArch to quit the game

Is there are workarounds but all of them are a lot less convenient. It really takes away from the handheld console experience when you have to start worrying about whether you closed your previous game a specific way.

**Comment by Jetup13** [comment]
> Why don't I have to do this on my other devices then? What you're suggesting is not a solution to this bug.

Two devices with Android 13's might not work the same. Especially if the vendor fails to properly configure it. This goes with any device or any Android software. Retroid Pocket 4 Pro has Daijishou properly closes RetroArch. But the Retroid Pocket 5 does not. Both are Android 13.

Another (different) example was the Odin 2. When it came out it was a nightmare for frontends. It had various issues such failing to properly open App's via am start argument or even not loading certain file extensions. This was fixed via OTA from what I gather.

Just use RetroArch to close RetroArch. ES-DE also recommends this. This is also a reason why a lot of people lose RetroArch save data or settings. Properly close RetroArch the way RetroArch wants you to close it. 



> > As I've said before, stop using the home button or app switcher and properly quit RetroArch beforehand. Using the home button or app switcher with RetroArch can possibly lead to save data or configuration loss as RetroArch might not get a chance to flush to disk. RetroArch is already a bit funky on Android. This is more of a android issue than a frontend issue.
> > RetroArch has many ways you can properly close it.
> > I'd suggest the following -
> > Enable Quit on Close Content:
> > > RetroArch > User Interface > Quit on Close Content > CLI > Save Configuration
> > 
> > 
> > Quick Menu:
> > > RetroArch Quick Menu > Close Content
> > 
> > 
> > Hotkeys:
> > > RetroArch > Settings > Hotkeys > Quit, Quit (Controller Combo), or Close Content if (Quit on Close Content is enabled) > Save Configuration
> > 
> > 
> > TLDR: Use RetroArch to quit the game
> 
> Is there are workarounds but all of them are a lot less convenient. It really takes away from the handheld console experience when you have to start worrying about whether you closed your previous game a specific way.

Using hotkeys might be the most convenient suggestion I listed. 

Unless someone is able to provide exact versions to where Daijishou is closing RetroArch or another frontend closing RetroArch then this is most likely a device vendor issue. This might be why not every device is experiencing this symptom.

**Comment by AgentSmith2518** [comment]
So I can confirm that this same issue happens on the Logitech G Cloud. Ive recently ran into the issue. What's odd is it didn't do it at first. But then I uninstalled Daijisho and RetroArch and started over again and did run into the issue.

**Comment by DrWafflePhD** [comment]
It's almost august and this is still a concern. Not only that, I can't set a hotkey to my home button, as retro arch seems to recognize the back button (which I have set to the quick menu) but not the home button.

**Comment by Pluckerpluck** [comment]
Just got a Retroid Pocket Flip 2, set it up, and I'm suffering from this.

It's annoying because when I'm playing arcade stuff I often really don't care about the sanctity of my save state or data. I just want to forcefully kill whatever I was doing and start a new game.

**Comment by mattakins** [comment]
This same issue occurs with Beacon and Console Launcher, but RetroidLauncher is able to kill existing instances of RetroArch and launch new games without first closing the previous one - just like how Daijishou used to work. Based on RetroidLauncher's GitHub, it uses the exact same launch arguments as Daijishou, so the issue isn't the launch arguments. It has to be with how the process killing is handled.

**Comment by mattakins** [comment]
My last comment was incorrect. I figured out how to fix this issue by editing the launch parameters for each player in Daijishou.
1. Go to Settings → Library → Scroll down to 'Inspect players'
2. Select a core you are using (double check if you're using RetroArch, RetroArch 32, or RetroArch 64)
3. Tap the 'Player am start arguments' box.
4. Scroll to the bottom and delete `--activity-clear-task` and `--activity-clear-top`.
5. Add `--activity-single-top`
6. Tap 'Next', then tap 'Done' in the next text box that pops up, then tap 'Save' in the bottom-right corner.
7. You can now launch games from Daijishou when RetroArch is already open without it crashing! 🎉

@TapiocaFox Would it make sense to make this the default launch parameter for all players?

**Comment by DrWafflePhD** [comment]
> My last comment was incorrect. I figured out how to fix this issue by editing the launch parameters for each player in Daijishou.
> 
> 1. Go to Settings → Library → Scroll down to 'Inspect players'
> 2. Select a core you are using (double check if you're using RetroArch, RetroArch 32, or RetroArch 64)
> 3. Tap the 'Player am start arguments' box.
> 4. Scroll to the bottom and delete `--activity-clear-task` and `--activity-clear-top`.
> 5. Add `--activity-single-top`
> 6. Tap 'Next', then tap 'Done' in the next text box that pops up, then tap 'Save' in the bottom-right corner.
> 7. You can now launch games from Daijishou when RetroArch is already open without it crashing! 🎉
> 
> [@TapiocaFox](https://github.com/TapiocaFox) Would it make sense to make this the default launch parameter for all players?

I did this, it allows Retro Arch to open again without crashing, but the game it opens is always the first game you opened until you close it manually. I would like if it could go back to killing the old game and opening the new game when hit from the launcher.

**Comment by mattakins** [comment]
Ahh ok - just tested this on the latest stable build and got the same result. I did some changes to how RetroArch handles Android launch intents in a test build. I guess it requires those fixes in RetroArch and not just the launch arguments. I'll attempt merge those changes into RetroArch. That paired with these new launch arguments will fix this issue.

**Comment by mspielberg** [comment]
@mattakins This sounds promising so I gave it a shot. Unfortunately it does not appear to be working as described. I installed the 2025-10-13 nightly of RetroArch on my A13 Retroid Mini, and adjusted the player arguments to include —activity-single-top.

Now when I launch RA, then hit Menu to return to Daijishou, and try to launch another game on the same core, I no longer get a black screen. Instead RetroArch exits entirely and dumps me back to Daijishou.

**Comment by mattakins** [comment]
@mspielberg Thanks for testing the fix. It has changed since my last comment. Please remove all other arguments aside from `--activity-clear-top`. This should work. Please report back. If it doesn't, can you please paste over the full launch command from Daijishou?

**Comment by mspielberg** [comment]
@mattakins 
The git commit of the RA build is 64ff381.
This is the full launch string:

```
-n com.retroarch/com.retroarch.browser.retroactivity.RetroActivityFuture
  -e ROM {file.path}
  -e LIBRETRO /data/data/com.retroarch/cores/mesen_libretro_android.so
  -e CONFIGFILE /storage/emulated/0/Android/data/com.retroarch/files/retroarch.cfg
  -e IME com.android.inputmethod.latin/.LatinIME
  -e DATADIR /data/data/com.retroarch
  -e APK /data/app/com.retroarch-1/base.apk
  -e SDCARD /storage/emulated/0
  -e EXTERNAL /storage/emulated/0/Android/data/com.retroarch/files
  --activity-clear-top
```

The behavior tested is launching a game, then pressing Home to return to Daijishou without exiting RA, then launching another game from Daijishou for the same platform. At which point RA exits without appearing. The play count etc. updates in Daijishou indicating it did attempt to launch the player. Trying again to launch at this point succeeds since RA is no longer running.

**Comment by mattakins** [comment]
@mspielberg Thanks for the report. I'll dig into this and report back.

**Comment by mattakins** [comment]
@mspielberg - I testd on my Retroid Pocket Mini Android 13 and --activity-clear-top is working for me. Can anyone else jump in and test? Be sure you have the right cores installed and troubleshoot all other variables.

@TapiocaFox - my fix has been accepted into the latest stable build of RetroArch. For all RetroArch players, if you eliminate --activity-clear-task and just leave only --activity-clear-top for the launch arguments this issue is totally resolved and you can launch as many games as you want over each other. Are you open to making this the default launch arguments so this will work out of the box? Thanks!

**Comment by Jetup13** [comment]
I have the fixes pending, just haven't pushed it yet. The problem is that is the following:

1. It completely kills the Google Play's version support (this is fine as users shouldn't be using that build)
2. The user must click an item in Daijishou twice to open the second game on some devices (this is bad as we'll get more support tickets)

I assume this behavior happens because RetroArch checks if a content is running, if yes kills it, but it doesn't proceed to load the next rom after like so

If you want to test the changes:
 > Daijishou > Download > Index URI > change URL to the following:

```https://raw.githubusercontent.com/Jetup13/DaijishouEXP/main/platforms/index.json```

**Comment by DrWafflePhD** [comment]
@mattakins 
I've just updated to the newest stable retroarch and changed the launch options. While it now allows me to open launch multiple retroarch games, it doesn't actually launch new games until the old game has be shutdown. 

So if I open Game A, it launches Game B. 
But if I click home and then open Game B, it still resumes Game A. 

I'd still like it to close Game A and launch B, like how it worked before.

**Comment by mattakins** [comment]
@DrWafflePhD I'm having trouble replicating this. What device and Android build are you on?
@Jetup13 Good to know - I think this double-click issue can be resolved. I'll look into it. Can you describe the Google Play version issue? I would think the functionality would be more dependent on the user having the latest RetroArch build.

Thanks everyone.

**Comment by Jetup13** [comment]
@mattakins 
Google Play version hasn't been updated in years and requires the massive am start. Here's an example:

```
-n com.retroarch/com.retroarch.browser.retroactivity.RetroActivityFuture 
-e ROM {file.path}
-e LIBRETRO /data/data/com.retroarch/cores/gpsp_libretro_android.so
-e CONFIGFILE /storage/emulated/0/Android/data/com.retroarch/files/retroarch.cfg
-e IME com.android.inputmethod.latin/.LatinIME
 -e DATADIR /data/data/com.retroarch
-e APK /data/app/com.retroarch-1/base.apk
-e SDCARD /storage/emulated/0
-e EXTERNAL /storage/emulated/0/Android/data/com.retroarch/files
--activity-clear-task
--activity-clear-top
```

While latest website stable/nightly just needs this:

```
-n com.retroarch/com.retroarch.browser.retroactivity.RetroActivityFuture
-e ROM {file.path}
-e LIBRETRO gpsp
-e CONFIGFILE /storage/emulated/0/Android/data/com.retroarch/files/retroarch.cfg
```

If we're going to update all of the am start commands might as well use the proper cleaner am start and direct people to use the website and nightly versions as Libretro recommends

Plus looking at some of the commita the past few months it looks like the Google Play version might eventually update

**Comment by DrWafflePhD** [comment]
@mattakins I'm on the Retroid Pocket 5, Android 13. Retroarch v1.21.0

**Comment by mattakins** [comment]
@Jetup13 This is definitely a lot cleaner. Would --activity-clear-top still come into play with the new am start?
@DrWafflePhD I will test on my Retroid Pocket 5.

**Comment by Jetup13** [comment]
It doesn't seem like we need it. I might add:

```--activity-no-animation```

That should remove the slide transitions when launching a game. Not sure if everyone likes that transitions

**Comment by mattakins** [comment]
@Jetup13 Hmm, not needing the flag is not what I would expect. I realized I'm using the Google Play build. Which Daijishou build are you using? The one here in the GitHub release?

EDIT: I was confused - you're referring to the nightly RetroArch build. My bad. I have enough info for testing now.

**Comment by Jetup13** [comment]
Google Play build of Daijishou. Both Google Play and Github versions are the same (minus a few things under-the-hood). Unless your talking about my message [above](https://github.com/TapiocaFox/Daijishou/issues/703#issuecomment-3712843727). For that I was talking about the RetroArch builds. Which everyone should be using the website or nightly versions

Other frontends also don't use the flags --activity-clear-task or --activity-clear-top for RetroArch (Example: [ES-DE](https://gitlab.com/es-de/emulationstation-de/-/blob/master/resources/systems/android/es_systems.xml?ref_type=heads#L745))

**Comment by mattakins** [comment]
@Jetup13 - Just submitted a fix for launching Game B with a single tap. This works great with your simplified launch argument. I vote to not include --activity-no-animation. I think the launch animations serve a purpose and create a sense of space within the operating system. I tested without it and it's a bit jarring. Also that flag sometimes results in a weird backward window movement animation.

@DrWafflePhD - Looks like you're on an earlier stable build of RetroARch. My original fix was incorporated in 1.22.0. This may also be your issue @mspielberg. Updating to the latest stable build should resolve, but you'll have to double tap to open a second game until this latest fix is incorporated. Or switch to nightly builds after this is merged.

**Comment by mspielberg** [comment]
@mattakins 

> This may also be your issue @mspielberg.

As I mentioned, I was running a nightly at githash 64ff381, which definitely included your fix:

https://github.com/libretro/RetroArch/compare/bd56674~1...64ff381

> Updating to the latest stable build should resolve, but you'll have to double tap to open a second game until this latest fix is incorporated. Or switch to nightly builds after this is merged.

I believe this is exactly the issue I was describing:

> Now when I launch RA, then hit Menu to return to Daijishou, and try to launch another game on the same core, I no longer get a black screen. Instead RetroArch exits entirely and dumps me back to Daijishou.

...

> The behavior tested is launching a game, then pressing Home to return to Daijishou without exiting RA, then launching another game from Daijishou for the same platform. At which point RA exits without appearing. The play count etc. updates in Daijishou indicating it did attempt to launch the player. Trying again to launch at this point succeeds since RA is no longer running.

So you need to double-tap to open a second game, because the first tap opens RA but causes it to immediately exit back to Daijishou. This happens so fast that it's a barely perceptible flicker.

I'm a little mystified about what changed between you not being able to reproduce the problem, and now having a fix for it exactly.

**Comment by mattakins** [comment]
@mspielberg Since it happens so fast, I didn't realize that RA was crashing until @Jetup13 described it as having to double-tap to launch a game. This was the issue you were experiencing. I was replicating it on my end without knowing it.

**Comment by lukevp** [comment]
@mattakins thanks for revisiting this issue and figuring out a way to make the launch work more reliably! how long does it typically take to get through the review process to nightly and eventually stable?

**Comment by mattakins** [comment]
@lukevp You're welcome! The fix was merged this morning so it should be in tonight's build. Stable releases are on average 4 - 6 months apart but this can vary. 1.22 came out in mid-November so could be some time before another release. Tagging @Jetup13 for visibility.

**Comment by Jetup13** [comment]
Seems like it's working. I'm going to go through the new configs a few times before I push the new am start arguments to make sure everything is correct. I don't want to screw up everyones platforms and create double entries 😅

**Comment by mattakins** [comment]
@Jetup13 Sounds great. Let me know if your run into anything else unexpected.

**Comment by mspielberg** [comment]
@mattakins I downloaded the latest RA nightly, and installed @Jetup13's Daijishou platform/player configs from https://github.com/TapiocaFox/Daijishou/issues/703#issuecomment-3712800762. Launching appears to work well, but I've noticed one significant gap: The normal RA shutdown / "on content closed" processing does not appear to be executed.

In particular, I have "Auto Save State" and "Auto Load State" turned on, so when RA exits it's supposed to save state for the active game, and reload it when I launch that game again. This works fine when I properly close RA via the menu or a configured hotkey.

However, with the fix, if I launch game A (let's call the state that RA loads at this time state A1), play it for a while, then go back to Daijishou and launch game B (let's call the state of game A that should be automatically saved at this time state A2), then game B launches fine with a single tap. But if I then exit game B and load game A again, the state that is loaded is state A1, and state A2 is lost. Looking in my file explorer, I can confirm that the `<game A>.state.auto` file is not modified when I launch game B.

At this point this seems to be more of an RA issue than a Daijishou issue. Should we move the discussion or continue here?

**Comment by mattakins** [comment]
@mspielberg Thanks for testing out the new build. Happy to continue the discussion here. What you're describing is expected behavior. I tested out my Android 11 Retroid Pocket 3+ (an Android version unaffected by this bug) with auto saving and auto loading states enabled in an older build of RetroArch before these changes. Loading a game when another game is open does not trigger auto save.

This fix restores simply restores this behavior and functionality in later Android versions. It's kind of like expecting RetroArch to auto save when force quitting. The best way to auto save is to go to Main Menu > Quit before launching a new game.

I also did a feasibility analysis, and implementing this would require introducing launch delays and significant changes to how subsequent launches are handled. The goal here is narrowly scoped: restore the ability to launch subsequent games when one is already open in the same way that was possible before recent Android changes broke it.

**Comment by Jetup13** [comment]
Little update, I went through 100+ platforms that contain retroarch players again. Again I haven't pushed it yet but probably will later today.

**Comment by Jetup13** [comment]
I pushed the player update. If you used the the test index url please reset it:

> Daijishou > Download > Index url > Reset > Save

Also please update to the latest RetroArch nightly build and test if its working correctly. If I does I can close up this issue

**Comment by mattakins** [comment]
@Jetup13 Great to hear. I'll test this tonight.

**Comment by mattakins** [comment]
@Jetup13 Just tested and it works great. Thanks!

**Comment by Jetup13** [comment]
Thanks. I think can finally close up this issue. If there's anything else don't hesitate to create another ticket
